### PR TITLE
[fix][broker] Fix calculate avg message per entry

### PR DESCRIPTION
**Pull request by coderzc**
### Motivation
Calculating avg message per entry using `entries.size()` is incorrect, because the entries may have `null` values.

### Modifications

Calculate avg message per entry using the number of filter-out Non-null value total entries.

### Documentation

Check the box below or label this PR directly.

Need to update docs? 

- [ ] `doc-required` 
(Your PR needs to update docs and you will update later)
  
- [x] `doc-not-needed` 
(Please explain why)
  
- [ ] `doc` 
(Your PR contains doc changes)

- [ ] `doc-complete`
(Docs have been already added)

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java [hunk 1]
```diff
@@ -290,13 +290,16 @@ public Future<Void> sendMessages(final List<? extends Entry> entries, EntryBatch
...
+        int totalEntries = 0;
+
+        for (int i = 0; i < entries.size(); i++) {
+            Entry entry = entries.get(i);
+            if (entry != null) {
+                totalEntries++;
+                // Note
+                // Must ensure that the message is written to the pendingAcks before sent is first,
+                // because this consumer is possible to disconnect at this time.
+                if (pendingAcks != null) {
```
Comment (Technoboy-) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java:302]: Why move `pendingAcks` here ?

Comment (coderzc) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java:302]: Make sure it can enter the loop.

**Review by Technoboy- — APPROVED** [review]

**Review by gaoran10 — APPROVED** [review]
LGTM, but I'm not sure why the entry could be null. Do you know the reason?

**Comment by coderzc** [comment]
> LGTM, but I'm not sure why the entry could be null. Do you know the reason?

Because some entries were set to `null` when the message was previously filtered.
Please see: https://github.com/apache/pulsar/blob/b1a29b520d34d60e60160e3a7b9b0e26926063ee/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/AbstractBaseDispatcher.java#L98

**Review by codelipenghui — COMMENTED** [review]
@coderzc The change looks good. Could you please help add a test? to avoid the regression in the future.

**Review by mattisonchao — APPROVED** [review]

**Comment by coderzc** [comment]
> @coderzc The change looks good. Could you please help add a test? to avoid the regression in the future.

ok, I have already added a test.

**Review by eolivelli — APPROVED** [review]
Lgtm

Good catch!

Please  cherry pick to 2.10 branch

**Comment by eolivelli** [comment]
I don't think we need this in 2.9, filters don't exist there

**Review by AnonHxy — APPROVED** [review]

**Comment by Technoboy-** [comment]
> I don't think we need this in 2.9, filters don't exist there

We need to cherry pick to 2.9 and 2.8,  the same issue in `sendMessages `

**Review by codelipenghui — APPROVED** [review]

**Comment by mattisonchao** [comment]
Hi, @coderzc 
Would you like to push a PR to branch-2.9 to fix this problem?

**Comment by coderzc** [comment]
> Hi, @coderzc Would you like to push a PR to branch-2.9 to fix this problem?

Yes, I will push a PR to branch-2.9.

**Comment by zymap** [comment]
Remove from release 2.8.5. Since the original code is introduced in 2.9.3.https://github.com/apache/pulsar/pull/14666

**Pull request merged**